File: src/core/scala/mrtjp/core/vec/Rect.scala
/*
 * Copyright (c) 2014.
 * Created by MrTJP.
 * All rights reserved.
 */
package mrtjp.core.vec

case class Rect(origin:Point, size:Size)
{
    def this(x:Int, y:Int, width:Int, height:Int) = this(Point(x, y), Size(width, height))
    def this(min:Point, max:Point) = this(min, Size(max.x-min.x, max.y-min.y))
    def this(r:Rect) = this(r.origin, r.size)

    override def equals(obj:scala.Any) = obj match
    {
        case that:Rect => origin == that.origin && size == that.size
        case _ => false
    }

    def copy = new Rect(origin, size)

    def x = origin.x
    def y = origin.y
    def width = size.width
    def height = size.height

    def maxX = x+width
    def maxY = y+height
    def maxPoint = Point(maxX, maxY)

    def midX = x+width/2
    def midY = y+height/2
    def midPoint = Point(midX, midY)

    def contains(p:Point):Boolean = p.x >= x && p.y >= y && p.x <= maxX && p.y <= maxY
    def contains(rect:Rect):Boolean = contains(rect.origin) && contains(rect.maxPoint)
    def intersects(r:Rect) = contains(r.origin) || contains(r.maxPoint) || r.contains(origin) || r.contains(maxPoint)

    def enclose(p:Point) = new Rect(Point(math.min(x, p.x), math.min(y, p.y)), Point(math.max(maxX, p.x), math.max(maxY, p.y)))
    def union(r:Rect) = new Rect(Point(math.min(x, r.x), math.min(y, r.y)), Point(math.max(maxX, r.maxX), math.max(maxY, r.maxY)))

    def ndc(p:Point):Vec2 = {
        val dx = ((p.x - origin.x) / width.toDouble * 2D) - 1
        val dy = ((p.y - origin.y) / height.toDouble * 2D) - 1
        Vec2(dx, -dy)
    }
}

object Rect
{
    val zeroRect = Rect(Point.zeroPoint, Size.zeroSize)
    val infiniteRect = Rect(-Point.infinitePoint/2, Size.infiniteSize)
}

File: core/src/main/java/mrtjp/projectred/lib/Rect.java
package mrtjp.projectred.lib;

public class Rect {

    public static final Rect ZERO = new Rect(Point.ZERO, Size.ZERO);
    public static final Rect INFINITE = new Rect(new Point(Integer.MIN_VALUE / 2, Integer.MIN_VALUE / 2), Size.INFINITE);

    public final Point origin;
    public final Size size;

    public Rect(Point origin, Size size) {
        this.origin = origin;
        this.size = size;
    }

    public Rect(int x, int y, int width, int height) {
        this(new Point(x, y), new Size(width, height));
    }

    public Rect(Point min, Point max) {
        this(min, new Size(max.x - min.x, max.y - min.y));
    }

    public int x() {
        return origin.x;
    }

    public int y() {
        return origin.y;
    }

    public int width() {
        return size.width;
    }

    public int height() {
        return size.height;
    }

    public int maxX() {
        return origin.x + size.width;
    }

    public int maxY() {
        return origin.y + size.height;
    }

    public Point maxPoint() {
        return new Point(maxX(), maxY());
    }

    public int midX() {
        return origin.x + size.width / 2;
    }

    public int midY() {
        return origin.y + size.height / 2;
    }

    public Point midPoint() {
        return new Point(midX(), midY());
    }

    public boolean contains(Point point) {
        return point.x >= origin.x && point.y >= origin.y && point.x <= maxX() && point.y <= maxY();
    }

    public boolean contains(Rect rect) {
        return contains(rect.origin) && contains(rect.maxPoint());
    }

    public boolean intersects(Rect rect) {
        return contains(rect.origin) || contains(rect.maxPoint()) || rect.contains(origin) || rect.contains(maxPoint());
    }

    public Rect enclose(Point p) {
        int x = Math.min(origin.x, p.x);
        int y = Math.min(origin.y, p.y);
        int w = Math.max(maxX(), p.x) - x;
        int h = Math.max(maxY(), p.y) - y;
        return new Rect(x, y, w, h);
    }

    public Rect union(Rect rect) {
        if (this.equals(ZERO)) return rect;
        if (rect.equals(ZERO)) return this;
        int x = Math.min(origin.x, rect.x());
        int y = Math.min(origin.y, rect.y());
        int w = Math.max(maxX(), rect.maxX()) - x;
        int h = Math.max(maxY(), rect.maxY()) - y;
        return new Rect(x, y, w, h);
    }

    public Rect expand(int x, int y) {
        int dw = width() + x < 0 ? -width() : x;
        int dh = height() + y < 0 ? -height() : y;
        return new Rect(origin.subtract(dw / 2, dh / 2), new Size(width() + dw, height() + dh));
    }

    public Rect trap(Rect rect) {
        int dx = (rect.x() < x() ? x() - rect.x() : 0) + (rect.maxX() > maxX() ? rect.maxX() - maxX() : 0);
        int dy = (rect.y() < y() ? y() - rect.y() : 0) + (rect.maxY() > maxY() ? rect.maxY() - maxY() : 0);
        return new Rect(rect.x() + dx, rect.y() + dy, rect.width(), rect.height());
    }

    public Vec2 ndc(Point p) {
        double dx = (2D * (p.x - origin.x) / width()) - 1;
        double dy = (2D * (p.y - origin.y) / height()) - 1;
        return new Vec2(dx, -dy);
    }
}
